[RFC] add support for integrating riscv-config to sail-riscv

liweiwei90 opened the pull request:
This is the initial version to integrating riscv-config to sail-riscv . It provides config2sail.py script to generate SAIL code from ISA YAML configuration. The generated code contains:  

- the structure of each csr register;
- the legalized function for each csr register(legalize_<csr_name>) and each WARL field in the register(legalize_<csr_name>_<field_name>)
- the total csr check function(is_csr_defined)          
- the total csr reset function (reset_regs)          
- the total csr read/write function (read_csr/write_csr)

We also modify the origin sail model to fit into the generated code, such as delete the original csr defines and modify the original csr read/write function. 
However, partial csr related functions still need additional support by handwriting, such as the read/write function of time register (It's partitioned into two registers in RV32, and will be generated as two register by our script, however they are used as a 64-bits register in the model). More cases can be found in the riscv_insns_zicsr.sail : readCSR/writeCSR function.

Any comments are welcome.

martinberger commented:
@liweiwei90 The PR doesn't currently build.

liweiwei90 commented:
Sorry. This PR need riscv-config to be installed before building sail-riscv.

scottj97 commented:
This is very difficult to review being one giant Git commit. Insignificant changes (like renaming variables `fflags` to `f_flags`) are mixed in with substantive changes, cluttering up the diff. There are no commit messages to explain the changes you made, either now during review or later during archaeology.

Please [learn](https://dustinspecker.com/posts/making-smaller-git-commits/) to [use](https://techstacker.com/make-small-git-commits/) [Git](https://bignerdranch.com/blog/small-distinct-commits-say-you-care/) [effectively](https://sethrobertson.github.io/GitBestPractices/) and you will have a much better chance of getting constructive reviews and your PRs merged.

scottj97 commented:
> Sorry. This PR need riscv-config to be installed before building sail-riscv.

The README needs to be updated to explain how to do this.

jrtc27 commented:
There's far too much gratuitous renaming going on here, much of which makes things worse (e.g. a bunch of named fields are now magic constant indices into a bit vector).

Various function have gone from taking typed arguments to xlenbits.

The DSL-within-a-DSL approach (a string expressing some kind of logical implication for legalisation, embedded inside YAML, used to generate the Sail) is pretty opaque to readers. Especially when it's being used for _architecturally-mandated_ behaviour like for pmpaddrX that really should not be changing between implementations; implementations are free to leave some of them unimplemented etc, but they cannot fail to respect the locking, which currently this PR puts in the _config_.

I disagree with fflags being split out from fcsr. It is not a separate register, it is a view into a subset of fcsr, therefore it should remain that in the model rather than duplicating the state. Removing it as a register will also mean you don't have to rename a bunch of fflags local variables.

The Python script doesn't conform to PEP-8, the standard Python style. Various linters, such as flake8, exist for this. Flags should be --like-this not --like_this. It also defaults to RV64; having a default sounds like a bad idea, it should always require it to be specified. It's also just quite hard to read; the parsing and generating are all done at once rather than separated out into two steps, and the parsing is a few big blocks of dense, uncommented code that can get up to 5 blocks deep.

Where are the actual CSR encodings now specified? How do I, as a downstream that implements a non-standard research extension that adds new CSRs, adapt my existing Sail model to keep supporting them? Currently all you need to do is define additional clauses for ext_read_CSR/ext_write_CSR/ext_is_CSR_defined.

Please go review your own diff as there are a bunch of obvious bogus diffs that come from blindly doing a find+replace of things "time" to "mtime", which catches both words that aren't referring to the CSR and substrings of words.

In general I'm also just concerned about the readability implications of having too much of this be autogenerated rather than a limited number of hooks for (a) if a CSR is implemented (b) if it is subject to legalisation that isn't governed by architectural state, what that behaviour is. When it's autogenerated, it's very difficult to put any amount of Sail into the privileged spec. Contrast this with the Armv8-A spec where many (most?) of the system registers have ASL that shows how what features you have and what the architectural is affect the behaviour of accessing that register. I realise RISC-V is less prescriptive than Armv8-A so having some of it behind a magic "implementation-defined legalisation" function is necessary, but we really should try to limit the generated code to just that.

liweiwei90 commented:
> This is very difficult to review being one giant Git commit. Insignificant changes (like renaming variables `fflags` to `f_flags`) are mixed in with substantive changes, cluttering up the diff. There are no commit messages to explain the changes you made, either now during review or later during archaeology. If a regression is later discovered, `git bisect` can do no better than pointing to this commit -- and good luck finding that bug in the 6500 lines of this diff.
> 
> Please [learn](https://dustinspecker.com/posts/making-smaller-git-commits/) to [use](https://techstacker.com/make-small-git-commits/) [Git](https://bignerdranch.com/blog/small-distinct-commits-say-you-care/) [effectively](https://sethrobertson.github.io/GitBestPractices/) and you will have a much better chance of getting constructive reviews and your PRs merged.



Sorry for this giant Git commit. I had thought to partition them. However, all the code modified seems to be needed to integrate the generated sail code by config2sail.py. If partitoned, they may be built failed before the lastest commit. Is that acceptable?

liweiwei90 commented:
> > Sorry. This PR need riscv-config to be installed before building sail-riscv.
> 
> The README needs to be updated to explain how to do this.

OK. I'll update this.

liweiwei90 commented:
> There's far too much gratuitous renaming going on here, much of which makes things worse (e.g. a bunch of named fields are now magic constant indices into a bit vector).
>
> Various function have gone from taking typed arguments to xlenbits.
>
> The DSL-within-a-DSL approach (a string expressing some kind of logical implication for legalisation, embedded inside YAML, used to generate the Sail) is pretty opaque to readers. Especially when it's being used for architecturally-mandated behaviour like for pmpaddrX that really should not be changing between implementations; implementations are free to leave some of them unimplemented etc, but they cannot fail to respect the locking, which currently this PR puts in the config. Same goes for xEPC legalisation (sort of, that one's a bit more special).

Thanks for your comments. To make the parsing progress simple, this initial version of script will generate similar related structure and functions for all csrs. As you said, some behaviour are architecturally-mandated. Maybe we should decide what is configurable? Or should this be guaranteed by riscv-config?

liweiwei90 commented:
> I disagree with fflags being split out from fcsr. It is not a separate register, it is a view into a subset of fcsr, therefore it should remain that in the model rather than duplicating the state. Removing it as a register will also mean you don't have to rename a bunch of fflags local variables.
This may need special handling. In current riscv-config  fflags and frm are not shadow fields of fcsr.

liweiwei90 commented:
> The Python script doesn't conform to PEP-8, the standard Python style. Various linters, such as flake8, exist for this. Flags should be --like-this not --like_this. It also defaults to RV64; having a default sounds like a bad idea, it should always require it to be specified. It's also just quite hard to read; the parsing and generating are all done at once rather than separated out into two steps, and the parsing is a few big blocks of dense, uncommented code that can get up to 5 blocks deep.

Sorry. I'll try to improve it later.

liweiwei90 commented:
>Where are the actual CSR encodings now specified? How do I, as a downstream that implements a non-standard research extension that adds new CSRs, adapt my existing Sail model to keep supporting them? Currently all you need to do is define additional clauses for ext_read_CSR/ext_write_CSR/ext_is_CSR_defined.

The CSR encodings is defined in riscv-config itself, such as the encoding for hpmcounter15 is defined in schemas/schemas_isa.yaml
`address: {type: integer, default: 0xC0F, allowed: [0xC0F]}  `
 and the final generated check yaml will contain the addresses for csrs.
I think it needs to update the riscv-config to add new CSRs.

liweiwei90 commented:
> Please go review your own diff as there are a bunch of obvious bogus diffs that come from blindly doing a find+replace of things "time" to "mtime", which catches both words that aren't referring to the CSR and substrings of words.

Sorry. I'll review and modify them.

liweiwei90 commented:
> In general I'm also just concerned about the readability implications of having too much of this be autogenerated rather than a limited number of hooks for (a) if a CSR is implemented (b) if it is subject to legalisation that isn't governed by architectural state, what that behaviour is. When it's autogenerated, it's very difficult to put any amount of Sail into the privileged spec. Contrast this with the Armv8-A spec where many (most?) of the system registers have ASL that shows how what features you have and what the architectural is affect the behaviour of accessing that register. I realise RISC-V is less prescriptive than Armv8-A so having some of it behind a magic "implementation-defined legalisation" function is necessary, but we really should try to limit the generated code to just that.

I think all csr related information can be found in the configure file. such as:
```
    misa:
        reset-val: 0x4014312D
        rv32:
            accessible: true
            mxl:
                implemented: true
                type:
                    warl:
                        dependency_fields: []
                        legal:
                          - mxl[1:0] in [0x1]
                        wr_illegal:
                          - unchanged
                description: Encodes the native base integer ISA width.
                shadow:
                shadow_type: rw
                msb: 31
                lsb: 30
            extensions:
                implemented: true
                type:
                    warl:
                        dependency_fields: []
                        legal:
                          - extensions[25:0] bitmask [0x014112D, 0x0000000]
                        wr_illegal:
                          - unchanged
                description: Encodes the presence of the standard extensions, with
                    a single bit per letter of the alphabet.
                shadow:
                shadow_type: rw
                msb: 25
                lsb: 0
            fields:
              - extensions
              - mxl
              -
                  -
                      - 26
                      - 29
        description: misa is a read-write register reporting the ISA supported by
            the hart.
        address: 769
        priv_mode: M
        rv64:
            accessible: false
```
We don't need to read the generated code.

PeterSewell commented:
On Sat, 20 Nov 2021 at 07:49, liweiwei90 ***@***.***> wrote:

> There's far too much gratuitous renaming going on here, much of which
> makes things worse (e.g. a bunch of named fields are now magic constant
> indices into a bit vector).
>
> Various function have gone from taking typed arguments to xlenbits.
>
> The DSL-within-a-DSL approach (a string expressing some kind of logical
> implication for legalisation, embedded inside YAML, used to generate the
> Sail) is pretty opaque to readers.
>
>
This is a fundamental question.  I've not been tracking this closely, so
perhaps it has been justified, but instead of _generating_ Sail, the more
natural thing would be to _parameterise_ the Sail model, internally, on the
relevant configuration choices.   Otherwise one does not have a single
RISC-V definition (that one can then read as a single artifact, and reason
about using a theorem prover when needed), but an enormous family of
definitions, and one can only really read and reason only about single
instances.

Peter



> Especially when it's being used for architecturally-mandated behaviour
> like for pmpaddrX that really should not be changing between
> implementations; implementations are free to leave some of them
> unimplemented etc, but they cannot fail to respect the locking, which
> currently this PR puts in the config. Same goes for xEPC legalisation (sort
> of, that one's a bit more special).
>
> Thanks for your comments. To make the parsing progress simple, this
> initial version of script will generate similar related structure and
> functions for all csrs. As you said, some behaviour are
> architecturally-mandated. Maybe we should decide what is configurable? Or
> should this be guaranteed by riscv-config?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/128#issuecomment-974611307>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABFMZZX4PTHWJD5L5VFUXYLUM5HDPANCNFSM5ILVJ52Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

scottj97 commented:
> Sorry for this giant Git commit. I had thought to partition them. However, all the code modified seems to be needed to integrate the generated sail code by config2sail.py. If partitoned, they may be built failed before the lastest commit. Is that acceptable?

No, broken commits are not useful. But I can imagine how this could be partitioned into many commits. Start with a config.yaml that matches Sail's existing implementation. Section by section, replace some of the handwritten code with generated code. If you need to rename variables along the way, those are separate commits. If the generated code will differ from the handwritten code, first make the edit by hand and commit to show that the alternative approach works. Then add the generation next. Everything is still working (for this one config) at every step of the way.

Once complete, you have a system that can support any config. The commit log can explain your thinking and reasoning for each change. If any regression is discovered later, `git bisect` can trace it back to the one small commit that is the culprit.

I'm not asking that you go back and redo this. But refactoring like this is a very useful skill. Martin Fowler's book Refactoring showed me the light and I am much more productive afterward.

scottj97 commented:
> The DSL-within-a-DSL approach (a string expressing some kind of logical implication for legalisation, embedded inside YAML, used to generate the Sail) is pretty opaque to readers. Especially when it's being used for _architecturally-mandated_ behaviour like for pmpaddrX that really should not be changing between implementations; implementations are free to leave some of them unimplemented etc, but they cannot fail to respect the locking, which currently this PR puts in the _config_. Same goes for xEPC legalisation (sort of, that one's a bit more special).

This is more a criticism of the riscv-config project than of this PR, and should probably be discussed there. But I agree that conflating implementation choices with architectural requirements in one "config" is concerning.

liweiwei90 commented:
> > Sorry for this giant Git commit. I had thought to partition them. However, all the code modified seems to be needed to integrate the generated sail code by config2sail.py. If partitoned, they may be built failed before the lastest commit. Is that acceptable?
> 
> No, broken commits are not useful. But I can imagine how this could be partitioned into many commits. Start with a config.yaml that matches Sail's existing implementation. Section by section, replace some of the handwritten code with generated code. If you need to rename variables along the way, those are separate commits. If the generated code will differ from the handwritten code, first make the edit by hand and commit to show that the alternative approach works. Then add the generation next. Everything is still working (for this one config) at every step of the way.
> 
> Once complete, you have a system that can support any config. The commit log can explain your thinking and reasoning for each change. If any regression is discovered later, `git bisect` can trace it back to the one small commit that is the culprit.
> 
> I'm not asking that you go back and redo this. But refactoring like this is a very useful skill. Martin Fowler's book Refactoring showed me the light and I am much more productive afterward.

Thanks for your suggestions. I'll try to update in this way in the future version.

allenjbaum commented:
This is quite an interesting discussion; there are lots of detailed issues, but two specific highlevel issues we need to address. The first is generated vs. parameterized Sail, and the other is whether risc-config should define both architectural requirements with implementation defined behavior.

These may be related, in that we might parameterize Sail for architectural requirements, but have generated Sail for the implementation . But first I'd like to see an example of what is architectural requirement vs. what is implementation choice.

I'm not seeing much in the riscv-config which I would say is an architectural requirement and NOT an implementation choice. Whether a CSR is implemented, whether it is implemented (so accesses don't trap) , or is implemented but read-only zero, or is set to zero on this trap but not that trap  -- those are all implementation choices. I'll have to look at what riscv-config says about PMP lock bits. As far as I know, they are allowed to be read-only zero or one for example (they weren't allowed to be readonly one in earlier spec versions.)

What a do see is a lot that can be easily parameterized, specifically anything whose implementation choice is binary: implemented or not, readonly or not, set to zero or not for this exception, for that exception... That works for a lot of optional behavior - maybe most of it. 

It doesn't work at all for others: xTVEC is WARL. Maybe the low 12 bits must be zero. Maybe the low 3 bits must be zero. Maybe it must be within a specific address range. Maybe it must be one of two specific values.  

On top of that, for WARL fields, the mapping from illegal to legal is completely arbitrary: leave it unchanged, set it to the smallest legal value, the nearest legal value, break any tie to be odd.even, etc, etc.  

All that is legal from a spec point of view, and trying to parameterize all of that in Sail will not be readable at all.
(the common case of leaving it unchanged could be a parameter, though)

I'm suggesting that we can do a lot of parameterization in Sail 
 - some via a generic function that can be shared by all CSRs, 
 - others that are specific to a CSR 
 - and the rest will need to be generated code. 
 So 3 specific methods

That is, unfortunately, a lot more work than having a script generate it all, because we have to go through all the possible CSRs and separate behavior that is global, those that are specific but easily described, and those that are difficult, and allow the latter to be generated code.

A slightly different, more readable way to to do this might be to define CSR specific functions that classify the  illegal values of a specific CSR field (or subfield) , and another that determines a legal value, as generated functions, and use those to ev

allenjbaum commented:
> > The DSL-within-a-DSL approach (a string expressing some kind of logical implication for legalisation, embedded inside YAML, used to generate the Sail) is pretty opaque to readers. Especially when it's being used for _architecturally-mandated_ behaviour like for pmpaddrX that really should not be changing between implementations; implementations are free to leave some of them unimplemented etc, but they cannot fail to respect the locking, which currently this PR puts in the _config_. Same goes for xEPC legalisation (sort of, that one's a bit more special).
> 
> This is more a criticism of the riscv-config project than of this PR, and should probably be discussed there. But I agree that conflating implementation choices with architectural requirements in one "config" is concerning.

I'll try to look into that more, but my impression wasn't that architecturally mandated behavior is not being covered by riscv-config. Architectural options, however, should be, and while locking behavior is architectural, whether or not a lock bit is readonly 0 or readonly 1 is an implementation choice

martinberger commented:
@liweiwei90 What's the status of this PR from your side?

liweiwei90 commented:
Sorry. I didn't take much time on this PR recently and in holiday now.  I’ll try the way suggested by Allen after our Spring Festival. Happy new year!

billmcspadden-riscv commented:
Hi folks.  I'm just ramping up on this,  so forgive me if I'm a bit naive with some of my questions/statements.
I have two items to discuss.

1)   I think I agree with what Peter says about parameterizations,  depending upon what "parameter" means.
In short,  I like run-time configurations (that is,  parameters in the code that tell the model how to behave
at *runtime*;  the parameters are passed in at invocation time)  and I dislike compile-time configurations
(things similar to #defines in C).  

2)  It appears we are trying to solve this problem using and parsing YAML.  I believe I know the historical
reason for this (the arch-test configuration is specified in YAML).   But the tech-config group has chosen
to adopt ASN.1 as the configuration "language" for RISC-V.    I think it bears discussing that we start with
that, rather than YAML.

3)  There are both opensource YAML and ANS.1 libraries (in C)  that allow you to parse the configurations file
based on a schema.  For my part,  I think these libraries should (could?) be brought into the RISC-V model.  And
the appropriate calls would be made to pull in the "config" file and then parameterize the model.

One more thing,  somewhat unrelated to this particular thread:  @liweiwei90 ,  may I please get your email address?
I would very much like to chat with you about this.

Bill Mc.

allenjbaum commented:
I think I agreed with point#1 (I hope I made that clear in the previous
email. )
I don't even think I really care if it's YAML or ASN.1, but the use cases
are different, and
it isn't so clear to me that there is full overlap between the
config-struct (ASN.1) and riscv-config (YAML), which is why I'm sticking
with YAML.

Specifically, I don't think there is support for some of the allowed WARL
functionality that we need, and
The use cases are different enough that there is no need in config-struct
to have the level of detail that riscv-config has for Sail.
There is nothing that impedes that, but no commitment to add it either.


What I am saying is that Sail and CSIM shouldn't have to care.
I'd like CSIM to have run-time command-line arguments that define internal
option variables as much as possible, so little or no parsing is required
at all.
A separate parser would read the configuration file and convert it to
something Sail can use at run time.
I don't really care too much whether that's ASN.1, or YAML, but we do have
some work invested in the YAML that I'd rather not throw away.
(There is a line item somewhere for the config-struct D.O.D to have a
translation tool between the two formats)

Most of those commmand-line arguments are pretty simple (e.g.
MISALIGN=TRUE, SATP_MODE=SV48_39, CSR_PMP_GRANULARITY=4096).
Slightly more complicated are typical CSRs (CSR_MISA_MASK=0x...,
CSR_MISA_VALUE=0x....)
More complicated are CSRs that have individual field values that have
different behaviors, or whose behavior has dependencies,
or fields with arbitrary subfields that can have different behaviors.
 (e.g. MTVEC where the low order address bits are either read/write
(optionally between some limits) or read_only_zero
  depending on the mode field, which has 1..3 specific values, and upper
address bits that are all zero)
So, something
like (CSR_MTVEC_ADDR="{CSR_MODE==1}[63:39]=0/[39:12]=0x800000..0x8000400%UNCHG/[11:0=0]"
                              (CSR_MTVEC_ADDR="{CSR_MODE==0}"....
or whatever it takes to make Sail parsing as simple as possible. IF that is
at all possible...

Separately: We need to establish the definitive list of architectural
options; it doesn't exist. There is a definitive list of ratified specs,
but that's only half the story.

On Mon, Feb 7, 2022 at 11:34 AM billmcspadden-riscv <
***@***.***> wrote:

> Hi folks. I'm just ramping up on this, so forgive me if I'm a bit naive
> with some of my questions/statements.
> I have two items to discuss.
>
>    1.
>
>    I think I agree with what Peter says about parameterizations,
>    depending upon what "parameter" means.
>    In short, I like run-time configurations (that is, parameters in the
>    code that tell the model how to behave
>    at *runtime*; the parameters are passed in at invocation time) and I
>    dislike compile-time configurations
>    (things similar to #defines in C).
>    2.
>
>    It appears we are trying to solve this problem using and parsing YAML.
>    I believe I know the historical
>    reason for this (the arch-test configuration is specified in YAML).
>    But the tech-config group has chosen
>    to adopt ASN.1 as the configuration "language" for RISC-V. I think it
>    bears discussing that we start with
>    that, rather than YAML.
>    3.
>
>    There are both opensource YAML and ANS.1 libraries (in C) that allow
>    you to parse the configurations file
>    based on a schema. For my part, I think these libraries should
>    (could?) be brought into the RISC-V model. And
>    the appropriate calls would be made to pull in the "config" file and
>    then parameterize the model.
>
> One more thing, somewhat unrelated to this particular thread: @liweiwei90
> <https://github.com/liweiwei90> , may I please get your email address?
> I would very much like to chat with you about this.
>
> Bill Mc.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/128#issuecomment-1031842218>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJTWP7B5VHPHPYGK2LLU2ANDVANCNFSM5ILVJ52Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

neelgala commented:
@billmcspadden-riscv 
I feel config-structure might not include the kind of granular information required to configure SAIL that maybe be easily available in the YAML - WARL being the primary feature.

The schemas for the YAML are present here : https://github.com/riscv-software-src/riscv-config/tree/master/riscv_config/schemas
I think it would be better to keep those schemas there instead of within SAIL - because riscv-config is used by several other tools as well. But if it really has to move, then riscv-config will have to pull them from the sail repo everytime.. and the schema in SAIL will have to be versioned (which can be the same as the root-level versioning of the repo) so that one can control which version of schema should be used by riscv-config.

@allenjbaum 
I don't know if mixing command-line AND having parsing based configuration is a good idea. For example, the warl string of satp.mode will capture what the next legal value is on an illegal-write. Now the command line will also define the set of legal values for satp.mode ? This will turn out to be quite chaotic don't you think where the YAML and the CLI options get out of sync - or did I get you wrong?

liweiwei90 commented:
It's pretty good to pass everything from command line and can use different config with the same sail model or csim. However with my current understanding, all the command line parameters are passed from c-emulator and ocaml-emulator to sail model, and the passed content are all bool or integer currently.  It's difficult to pass complicate config for me. Can we firstly pass simple parameters from command line with WARL related function as generated sail code?

allenjbaum commented:
My understanding is that the parameters are passed into the csim model, and
are used as variables inside i
That means the csim doesn't change based on those command-line options
(except perhaps RV32 vs. RV64? Not sure what happens if XLEN can be
read/write).

If we have different sail code for every CSR, then we no longer have a
golden model; we have a custom model for each implementation.
It makes it much more difficult to encapsulate the csim in a docker image -
maybe impossible
We need to have a single parameterized sim.
We don't have to pass the parameters on the command line, as opposed to
reading the YAML file directly, or piped through a pre-parser if that
simplifies things.
Those parameters, after parsing, could be strings, they don't have to be
integers or bools.
I have a proposed parsed format which consists primarily of lists of
integers - which can't be represented as integers,
but Peter is suggesting that the parsing could be done entirely inside the
Sail model, directly from the YAML.

On Tue, Feb 15, 2022 at 5:53 PM liweiwei90 ***@***.***> wrote:

> It's pretty good to pass everything from command line and can use
> different config with the same sail model or csim. However with my current
> understanding, all the command line parameters are passed from c-emulator
> and ocaml-emulator to sail model, and the passed content are all bool or
> integer currently. It's difficult to pass complicate config for me. Can we
> firstly pass simple parameters from command line with WARL related function
> as generated sail code?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/128#issuecomment-1041010638>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJWT2KD5MSCS6IJ45D3U3L7S5ANCNFSM5ILVJ52Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

neelgala commented:
> but Peter is suggesting that the parsing could be done entirely inside the
Sail model, directly from the YAML.

Parsing within the SAIL model would be ideal. Even better would be to take the YAML as a CLI input and then parse it - avoids rebuilding the SAIL model everytime the YAML changes. Will make things much simpler for riscof as well.

allenjbaum commented:
I'm still thinking that pre-parsing it into a format easier for sail to
digest, and passing that as a CLI (or just letting Sail read the
pre-parsed data) might simplify the task.
There would be less Sail code, and it would be easier to read.
The pre-parsed input could be sorted to make everything easier as well.
But if not, just point Sail at the YAML file.

OR... are we really passing that to the csim, not to the Sail?
I was told most of CLI are just passed on to the Cmodel, which uses them as
conditions of which code to execute.

On Tue, Feb 15, 2022 at 10:52 PM Neel Gala ***@***.***> wrote:

> but Peter is suggesting that the parsing could be done entirely inside the
> Sail model, directly from the YAML.
>
> Parsing within the SAIL model would be ideal. Even better would be to take
> the YAML as a CLI input and then parse it - avoids rebuilding the SAIL
> model everytime the YAML changes. Will make things much simpler for riscof
> as well.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/128#issuecomment-1041171936>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AHPXVJRLQOFSOOSRS5IZBPLU3NCTFANCNFSM5ILVJ52Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

PeterSewell commented:
On Wed, 16 Feb 2022 at 05:54, Allen Baum ***@***.***> wrote:

> My understanding is that the parameters are passed into the csim model, and
> are used as variables inside i
> That means the csim doesn't change based on those command-line options
>

y


> (except perhaps RV32 vs. RV64? Not sure what happens if XLEN can be
> read/write).
>

IIRC the model doesn't support dynamic change of XLEN


>
> If we have different sail code for every CSR, then we no longer have a
> golden model; we have a custom model for each implementation.
> It makes it much more difficult to encapsulate the csim in a docker image -
> maybe impossible
> We need to have a single parameterized sim.
>

y

(though you could, if you really have to, have a single model that includes
e.g.
a list of exotic legalisation functions that have been needed for previous
impls)


> We don't have to pass the parameters on the command line, as opposed to
> reading the YAML file directly, or piped through a pre-parser if that
> simplifies things.
> Those parameters, after parsing, could be strings, they don't have to be
> integers or bools.
> I have a proposed parsed format which consists primarily of lists of
> integers - which can't be represented as integers,
>


> but Peter is suggesting that the parsing could be done entirely inside the
> Sail model, directly from the YAML.
>

I wasn't saying anything about how the parsing should be done, only about
how the result should be represented within Sail.


>
> On Tue, Feb 15, 2022 at 5:53 PM liweiwei90 ***@***.***> wrote:
>
> > It's pretty good to pass everything from command line and can use
> > different config with the same sail model or csim. However with my
> current
> > understanding, all the command line parameters are passed from c-emulator
> > and ocaml-emulator to sail model, and the passed content are all bool or
> > integer currently. It's difficult to pass complicate config for me. Can
> we
> > firstly pass simple parameters from command line with WARL related
> function
> > as generated sail code?
> >
> > —
> > Reply to this email directly, view it on GitHub
> > <https://github.com/riscv/sail-riscv/pull/128#issuecomment-1041010638>,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AHPXVJWT2KD5MSCS6IJ45D3U3L7S5ANCNFSM5ILVJ52Q
> >
> > .
> > Triage notifications on the go with GitHub Mobile for iOS
> > <
> https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675
> >
> > or Android
> > <
> https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub
> >.
> >
> > You are receiving this because you were mentioned.Message ID:
> > ***@***.***>
> >
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/128#issuecomment-1041135772>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABFMZZTHWJGPH5ENKBFK3JTU3M32BANCNFSM5ILVJ52Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

allenjbaum commented:
Not having a dynamic XLEN is #2 on the list of "remedial" updates needed to support v1.1 of the architectural spec.
It does sound rather challenging. I'm hoping we won't ever need to implement it, but at some point someone will want to produce a chip that can run both RV64 and RV32 binaries - and as a guest under a hypervisor- and then we'll have a lot of work to do.

Having a list of exotic legalization functions can "work", for a very large value (and amount) of "work". The logistics sounds nasty.  I'm voting for putting the work up front so we don't have to deal with that.

I clearly misinterpreted it your comment - more than once probably.
What you've shown appears to partially match my thinking (though I'm having trouble interpreting the syntax, since I am not at all proficient in Sail), but doesn't cover all the cases that we need to cover, specifically the conditional cases, and the common  single bit WARL cases (where each bit can be either 0,1, or RW). I don't know if it handles the arbitrary subfield widths, though.
Otherwise, the others shouldn't be hard to add to this (by someone who knows Sail, anyway). 

The parsing part, and who those parsed values are incorporated into the Sail model, are still the open questions.

allenjbaum commented:
There are currently 3 proposals on the table for handling Sail configuration via the riscv-config YAML file:
a. This pull request: this uses a Python script to produce Sail code for the read/write functions for each CSR.
  a.1.: - this means it's a different Sail model for every implementation, and that we can't ship a single golden csim model inside a docker container, it has to be recompiled every time it is used. 
  a.2: - it isn't clear how readable the code will be.
  a.3 + it uses the latest riscv-config schema

b. The rv-config branch: This uses a separate Ocaml program that parses the YAML and produces, for each CSR, produces a Sail type which can be used to paramterize the write function.
  b.1 -This is superficially like the above; I don't know that a single precompiled csim could be shipped
  b.2 -The parser is written in Ocaml, which may be powerful, but is not well known and we'd have maintenance problem 
           (to be fair, you could say that about Sail, too...)
  b.3 -It uses an older version of the riscv-config schema and would need a major upgrade

c: My proposal: this would use a python or c script that parses the YAML into very easy to decode per/CSR lists that are passed into Sail as configuration values. Sail would parse each element into XLEN width bitfields and LOG2XLEN width integers (and/or lists of them). The Sail CSR write function would treat each element (or element of a list of elements) as a rule and apply it to the CSR and the write value sequentially until the list is consumed .
  c.1 - the parser doesn't exist yet
  c.2  - the sail code doesn't exist yet (though it would resemble the rv-config branch support.
  c.3  + no new types are needed, and only a single csim is generated

Note that all 3 proposals do the bulk of their parsing in a separate bit of (non-sail) code; the primary difference is whether individual Sail functions per CSR are generated, individual Sail types per CSR are generated, or text strings per CSR are generated. 

One unknown to me is whether making changes enabling a single csim will affect other uses of the formal model; that is, does the "data dependency" of using a string (or any other basic data type) to legalize CSR writes mean that theorem proving becomes impossible? Does it affect the OCaml model?

billmcspadden-riscv commented:
From my perspective,  option c is what we want to do.    Note:  there are C
libraries for
reading in YAML files.  And there are also libraries for reading ASN.1
specifications.

That's my 2 cents.

Bill Mc.


On Fri, Feb 18, 2022 at 9:26 PM Allen Baum ***@***.***> wrote:

> There are currently 3 proposals on the table for handling Sail
> configuration via the riscv-config YAML file:
> a. This pull request: this uses a Python script to produce Sail code for
> the read/write functions for each CSR.
> a.1.: - this means it's a different Sail model for every implementation,
> and that we can't ship a single golden csim model inside a docker
> container, it has to be recompiled every time it is used.
> a.2: - it isn't clear how readable the code will be.
> a.3 + it uses the latest riscv-config schema
>
> b. The rv-config branch: This uses a separate Ocaml program that parses
> the YAML and produces, for each CSR, produces a Sail type which can be used
> to paramterize the write function.
> b.1 -This is superficially like the above; I don't know that a single
> precompiled csim could be shipped
> b.2 -The parser is written in Ocaml, which may be powerful, but is not
> well known and we'd have maintenance problem
> (to be fair, you could say that about Sail, too...)
> b.3 -It uses an older version of the riscv-config schema and would need a
> major upgrade
>
> c: My proposal: this would use a python or c script that parses the YAML
> into very easy to decode per/CSR lists that are passed into Sail as
> configuration values. Sail would parse each element into XLEN width
> bitfields and LOG2XLEN width integers (and/or lists of them). The Sail CSR
> write function would treat each element (or element of a list of elements)
> as a rule and apply it to the CSR and the write value sequentially until
> the list is consumed .
> c.1 - the parser doesn't exist yet
> c.2 - the sail code doesn't exist yet (though it would resemble the
> rv-config branch support.
> c.3 + no new types are needed, and only a single csim is generated
>
> Note that all 3 proposals do the bulk of their parsing in a separate bit
> of (non-sail) code; the primary difference is whether individual Sail
> functions per CSR are generated, individual Sail types per CSR are
> generated, or text strings per CSR are generated.
>
> One unknown to me is whether making changes enabling a single csim will
> affect other uses of the formal model; that is, does the "data dependency"
> of using a string (or any other basic data type) to legalize CSR writes
> mean that theorem proving becomes impossible? Does it affect the OCaml
> model?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/riscv/sail-riscv/pull/128#issuecomment-1045655595>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AXROLOAZEU3MZBFBRTKLOQLU34EWNANCNFSM5ILVJ52Q>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Bill McSpadden
Formal Verification Engineer
RISC-V International
mobile:  503-807-9309

allenjbaum commented:
I am still concerned about how that affects the theorem proving properties of the models.
Any thoughts about that, Peter?

ptomsich commented:
Folks, we have no activity on this one for 15 months (and it doesn't even build at the moment): should we consider this PR as abandoned (and close it as such)?

ptomsich commented:
Closing as abandoned.
If you don't agree, please reopen with an updated changeset addressing any outstanding issues.
